apps/api/views/chat.py

```diff
@@ -55,13 +62,63 @@ def check_experiment_access(experiment, participant_id):
     return None
 
 
-def check_session_access(session):
+def handle_embedded_widget_auth(request, experiment_id=None, session=None):
+    embed_key = request.headers.get("X-Embed-Key")
+    if not embed_key:
+        return False, None, None
+
+    # Extract origin domain from headers
+    origin_domain = extract_domain_from_headers(request)
+    if not origin_domain:
+        raise PermissionDenied("Origin or Referer header required for embedded widgets")
+
+    if experiment_id:
+        is_valid, experiment_channel = validate_embed_key_for_experiment(
+            token=embed_key, origin_domain=origin_domain, experiment_id=experiment_id
+        )
+    elif session:
+        is_valid, experiment_channel = validate_embedded_widget_request(
+            token=embed_key, origin_domain=origin_domain, team=session.team
+        )
+        if is_valid and session.experiment_channel.platform != ChannelPlatform.EMBEDDED_WIDGET:
+            raise PermissionDenied("Session does not belong to an embedded widget channel")
+    else:
+        raise ValueError("Either experiment_id or session must be provided")
+
+    if not is_valid:
+        raise PermissionDenied("Invalid embed key or domain not allowed")
+
+    # Generate id for anon widget users
+    participant_remote_id = None
+    if hasattr(request, "data") and request.data:
+        participant_remote_id = request.data.get("participant_remote_id")
+
+    if not participant_remote_id:
+        participant_remote_id = f"embed_{uuid.uuid4()}"
+
+    return True, experiment_channel, participant_remote_id
+
+
+def check_session_access(session, request=None):
     """
-    Check if the request has access to the session based on experiment's public API settings.
+    Check if the request has access to the session.
+    Now handles both authenticated users and embedded widgets.
 
     Returns:
         Response object if access denied, None if access allowed
     """
+    if session.experiment_channel.platform == ChannelPlatform.EMBEDDED_WIDGET:
```

Comment:
We might want to prefetch the experiment channel for all callers of `check_session_access`